services/teamcity/teamcity-build.service.js

```diff
@@ -1,124 +1,115 @@
 'use strict'
 
-const LegacyService = require('../legacy-service')
-const { makeBadgeData: getBadgeData } = require('../../lib/badge-data')
+const Joi = require('joi')
+const TeamCityBase = require('./teamcity-base')
 
-// This legacy service should be rewritten to use e.g. BaseJsonService.
-//
-// Tips for rewriting:
-// https://github.com/badges/shields/blob/master/doc/rewriting-services.md
-//
-// Do not base new services on this code.
-function teamcityBadge(
-  url,
-  buildId,
-  advanced,
-  format,
-  data,
-  sendBadge,
-  request
-) {
-  const apiUrl = `${url}/app/rest/builds/buildType:(id:${buildId})?guest=1`
-  const badgeData = getBadgeData('build', data)
-  request(
-    apiUrl,
-    { headers: { Accept: 'application/json' } },
-    (err, res, buffer) => {
-      if (err != null) {
-        badgeData.text[1] = 'inaccessible'
-        sendBadge(format, badgeData)
-        return
-      }
-      try {
-        const data = JSON.parse(buffer)
-        if (advanced)
-          badgeData.text[1] = (
-            data.statusText ||
-            data.status ||
-            ''
-          ).toLowerCase()
-        else badgeData.text[1] = (data.status || '').toLowerCase()
-        if (data.status === 'SUCCESS') {
-          badgeData.colorscheme = 'brightgreen'
-          badgeData.text[1] = 'passing'
-        } else {
-          badgeData.colorscheme = 'red'
-        }
-        sendBadge(format, badgeData)
-      } catch (e) {
-        badgeData.text[1] = 'invalid'
-        sendBadge(format, badgeData)
-      }
+const buildStatusRegex = /SUCCESS|FAILURE|ERROR/
```

Comment:
The error messages will be a little clearer if you use `Joi.equal('SUCCESS', 'FAILURE', 'ERROR')`. (Same in the tests.)

Reply:
👍 I learn more about Joi's capabilities every day!